### PR TITLE
feat(cc): add value event and notifications for Multilevel Switch CC Set and Start/StopLevelChange

### LINKED ISSUE #4064: Add value notification event for Multilevel Switch CC
**Is your feature request related to a problem? Please describe.**
The Aeotec ZW111 is an in-wall dimmer with a single output but wiring for 2 external switches. It can be configured such that the second switch only sends zwave commands to an association group rather than controlling local load. It supports sending 2 sets of commands: basic and multilevel switch CC.

Currently, we can take advantage of this by associating the device with the controller node and setting it to send only basic commands - these are received by zwave-js and converted to value notifications. However, this doesn't work when set to multievel switch CC. 

If this were supported, then not only could you program automations (eg. in Home Assistant) for on/off events, but also for multilevel start/stop commands which correspond to holding and releasing the external switch.

**Describe the solution you'd like**
1. Support parsing of MultilevelSwitchCCSet and MultilevelSwitchCCStartLevelChange (MultilevelSwitchCCStop already supported as there are no options)
2. Support generating value notifications when the controller receives these commands. 

**Describe alternatives you've considered**
N/A

**Additional context**
Log from when I toggled then held-and-released external switch that had direct association with the controller and was set to send Multilevel Switch CCs
```
Subscribed to Z-Wave JS Log Messages…
2022-01-08T22:45:32.030Z SERIAL « 0x010c0004000d042601ff01da00fd (14 bytes)
2022-01-08T22:45:32.030Z DRIVER Dropping message because it could not be deserialized: MultilevelSwitchCCSet: 
                                  deserialization not implemented (ZW0320)
2022-01-08T22:45:32.031Z SERIAL » [ACK] (0x06)
2022-01-08T22:45:34.329Z SERIAL « 0x010c0004000d0426010001da0002 (14 bytes)
2022-01-08T22:45:34.330Z DRIVER Dropping message because it could not be deserialized: MultilevelSwitchCCSet: 
                                  deserialization not implemented (ZW0320)
2022-01-08T22:45:34.331Z SERIAL » [ACK] (0x06)
2022-01-08T22:45:37.229Z SERIAL « 0x010d0004000d052604200001d80025 (15 bytes)
2022-01-08T22:45:37.230Z DRIVER Dropping message because it could not be deserialized: MultilevelSwitchCCStart
                                  LevelChange: deserialization not implemented (ZW0320)
2022-01-08T22:45:37.231Z SERIAL » [ACK] (0x06)
2022-01-08T22:45:37.829Z SERIAL « 0x010a0004000d022605d80005 (12 bytes)
2022-01-08T22:45:37.830Z SERIAL » [ACK] (0x06)
2022-01-08T22:45:37.831Z DRIVER « [Node 013] [REQ] [ApplicationCommand]
                                  └─[MultilevelSwitchCCStopLevelChange]
2022-01-08T22:45:37.833Z CNTRLR « [Node 013] TODO: no handler for application command
```

### PR DESCRIPTION
**Pull request by reubenbijl**
Hi,

Would really appreciate some feedback on this   so I can keep working on it.

fixes: #4064

### CODE REVIEW

**Comment by reubenbijl** [comment]
@zwave-js-bot fix lint

**Comment by AlCalzone** [comment]
This doesn't pass the linter:
```
/home/runner/work/node-zwave-js/node-zwave-js/packages/zwave-js/src/lib/node/Node.ts
Error:   2826:9  error  'sourceEndpoint' is assigned a value but never used  @typescript-eslint/no-unused-vars
```

and tests are failing too - you can run them locally using `yarn test` to figure out what's wrong.

File: packages/zwave-js/src/lib/commandclass/MultilevelSwitchCC.ts [hunk 1]
```diff
@@ -497,6 +506,17 @@ export class MultilevelSwitchCC extends CommandClass {
 
 		await this.refreshValues();
 
+		// create compat event value if necessary
+		if (node.deviceConfig?.compat?.treatBasicSetAsEvent) {
```
Comment (AlCalzone) [inline packages/zwave-js/src/lib/commandclass/MultilevelSwitchCC.ts]: MultilevelSwitchCC should have a separate compat flag (not named `...Basic...`)

File: packages/zwave-js/src/lib/commandclass/MultilevelSwitchCC.ts [hunk 2]
```diff
@@ -581,24 +601,49 @@ export class MultilevelSwitchCCSet extends MultilevelSwitchCC {
...
+		label: "Target value",
+	})
+	public readonly targetValue: number | undefined;
+
+	@ccValue({ minVersion: 2 })
+	@ccValueMetadata({
+		...ValueMetadata.ReadOnlyDuration,
+		label: "Remaining duration until target value",
+	})
+	public readonly duration: Duration | undefined;
```
Comment (AlCalzone) [inline packages/zwave-js/src/lib/commandclass/MultilevelSwitchCC.ts]: This seems wrong. A `MultilevelSwitchCCSet` doesn't have a `currentValue` property and its properties shouldn't be readonly.

File: packages/zwave-js/src/lib/commandclass/MultilevelSwitchCC.ts [hunk 3]
```diff
@@ -581,24 +601,49 @@ export class MultilevelSwitchCCSet extends MultilevelSwitchCC {
...
 	public serialize(): Buffer {
-		const payload = [this.targetValue];
-		if (this.version >= 2 && this.duration) {
-			payload.push(this.duration.serializeSet());
+		const payload: number[] = [
+			typeof this.currentValue !== "number" ? 0xfe : this.currentValue,
+		];
+		if (this.version >= 2 && this.targetValue && this.duration) {
+			payload.push(this.targetValue, this.duration.serializeReport());
 		}
```
Comment (AlCalzone) [inline packages/zwave-js/src/lib/commandclass/MultilevelSwitchCC.ts]: Changing the code that sends `MultilevelSwitchCCSet` isn't necessary to understand them when they are received.

File: packages/zwave-js/src/lib/node/Node.ts [hunk 4]
```diff
@@ -2811,6 +2820,57 @@ protocol version:      ${this._protocolVersion}`;
...
+			});
+			this._valueDB.setValue(
+				getMultilevelSwitchCCCompatEventValueId(command.endpointIndex),
+				"stop",
+				{
+					stateful: false,
+				},
+			);
+			return;
+		}
```
Comment (AlCalzone) [inline packages/zwave-js/src/lib/node/Node.ts]: Let's not water down the meaning of the event value too much, it should be limited to the value received via a Set command.
I'd prefer having start/stopLevelChange as a new `notification` type, like for Powerlevel CC and Entry Control CC - see point 3. in my comment in the issue.

File: packages/zwave-js/src/lib/node/Node.ts [hunk 5]
```diff
@@ -2811,6 +2820,57 @@ protocol version:      ${this._protocolVersion}`;
 		}
 	}
 
+	/** Handles the receipt of a MultilevelCC Set or Report */
+	private handleMultilevelSwitchCommand(command: MultilevelSwitchCC): void {
+		// Retrieve the endpoint the command is coming from
+		const sourceEndpoint =
+			this.getEndpoint(command.endpointIndex ?? 0) ?? this;
```
Comment (AlCalzone) [inline packages/zwave-js/src/lib/node/Node.ts]: This isn't used

File: packages/zwave-js/src/lib/node/Node.ts [hunk 6]
```diff
@@ -2239,6 +2246,8 @@ protocol version:      ${this._protocolVersion}`;
 
 		if (command instanceof BasicCC) {
 			return this.handleBasicCommand(command);
+		} else if (command instanceof MultilevelSwitchCC) {
+			return this.handleMultilevelSwitchCommand(command);
```
Comment (AlCalzone) [inline packages/zwave-js/src/lib/node/Node.ts:2287]: That's fine

**Review by AlCalzone — CHANGES_REQUESTED** [review]

File: packages/zwave-js/src/lib/commandclass/MultilevelSwitchCC.ts (hunk 3 above)
Comment (AlCalzone) [reply packages/zwave-js/src/lib/commandclass/MultilevelSwitchCC.ts]: This is probably why tests are failing.

**Comment by reubenbijl** [comment]
@AlCalzone  Thanks for the feedback, I believe that I've made most of the changes you've suggested. I'd be keen if you have any thoughts on the event structure, with the way I'm passing the event type through the notification.

File: maintenance/schemas/device-config.json [hunk 7]
```diff
@@ -282,6 +282,9 @@
 				"treatBasicSetAsEvent": {
 					"const": true
 				},
+				"treatMultilevelSwitchAsEvent": {
```
Comment (AlCalzone) [inline maintenance/schemas/device-config.json]:
```suggestion
				"treatMultilevelSwitchSetAsEvent": {
```
(rename the other occurences similarly)

File: packages/zwave-js/src/lib/commandclass/MultilevelSwitchCC.ts [hunk 8]
```diff
@@ -109,6 +111,33 @@ function getSuperviseStartStopLevelChangeValueId(): ValueID {
...
+		property: "event",
+	};
+}
+
+/**
+ * @publicAPI
+ * This is emitted when a start or stop event is received
+ */
+export interface ZWaveNotificationCallbackArgs_MultilevelSwitchCC {
+	eventType: MultilevelSwitchCommand;
```
Comment (AlCalzone) [inline packages/zwave-js/src/lib/commandclass/MultilevelSwitchCC.ts]:
```suggestion
	eventType: MultilevelSwitchCommand.StartLevelChange | MultilevelSwitchCommand.StopLevelChange;
```
would be more specific

File: packages/zwave-js/src/lib/commandclass/MultilevelSwitchCC.ts [hunk 9]
```diff
@@ -581,22 +621,26 @@ export class MultilevelSwitchCCSet extends MultilevelSwitchCC {
...
+				this.duration = Duration.parseReport(this.payload[1]);
+			}
 		} else {
 			this.targetValue = options.targetValue;
 			this.duration = options.duration;
 		}
 	}
 
-	public targetValue: number;
+	public targetValue: number | undefined;
```
Comment (AlCalzone) [inline packages/zwave-js/src/lib/commandclass/MultilevelSwitchCC.ts]: `targetValue` in a MultilevelSwitchCCSet is always defined.

File: packages/zwave-js/src/lib/commandclass/MultilevelSwitchCC.ts [hunk 10]
```diff
@@ -581,22 +621,26 @@ export class MultilevelSwitchCCSet extends MultilevelSwitchCC {
...
 	public duration: Duration | undefined;
 
 	public serialize(): Buffer {
-		const payload = [this.targetValue];
+		const payload = [
+			typeof this.targetValue !== "number" ? 0xfe : this.targetValue,
+		];
```
Comment (AlCalzone) [inline packages/zwave-js/src/lib/commandclass/MultilevelSwitchCC.ts]: so this is not necessary (and doesn't make sense)

File: packages/zwave-js/src/lib/commandclass/MultilevelSwitchCC.ts [hunk 11]
```diff
@@ -581,22 +621,26 @@ export class MultilevelSwitchCCSet extends MultilevelSwitchCC {
...
-			// TODO: Deserialize payload
-			throw new ZWaveError(
-				`${this.constructor.name}: deserialization not implemented`,
-				ZWaveErrorCodes.Deserialization_NotImplemented,
-			);
+			validatePayload(this.payload.length >= 1);
+			this.targetValue = parseNumber(this.payload[0]);
+
+			if (this.version >= 2 && this.payload.length >= 2) {
+				this.targetValue = parseNumber(this.payload[0]);
```
Comment (AlCalzone) [inline packages/zwave-js/src/lib/commandclass/MultilevelSwitchCC.ts]:
```suggestion
			if (this.payload.length >= 2) {
```

Comment (AlCalzone) [reply packages/zwave-js/src/lib/commandclass/MultilevelSwitchCC.ts]: Only the duration is optional and it is guaranteed to be at index 1 if the payload is long enough.

File: packages/zwave-js/src/lib/commandclass/MultilevelSwitchCC.ts [hunk 12]
```diff
@@ -705,19 +749,27 @@ export class MultilevelSwitchCCStartLevelChange extends MultilevelSwitchCC {
...
 		} else {
 			this.duration = options.duration;
 			this.ignoreStartLevel = options.ignoreStartLevel;
 			this.startLevel = options.startLevel ?? 0;
 			this.direction = options.direction;
+			this.eventType = MultilevelSwitchCommand.StartLevelChange;
 		}
 	}
 
+	public eventType: MultilevelSwitchCommand;
```
Comment (AlCalzone) [inline packages/zwave-js/src/lib/commandclass/MultilevelSwitchCC.ts]: The class instance already identifies this command, no need for an extra variable. Just handle this like you handle the StopLevelChange command.

File: packages/zwave-js/src/lib/node/Node.ts [hunk 13]
```diff
@@ -2847,6 +2857,53 @@ protocol version:      ${this.protocolVersion}`;
...
+				endpoint: command.endpointIndex,
+				message:
+					"treating MultilevelSwitchCCStartLevelChange::Set as a value event",
+			});
+			// Notify listeners
+			this.emit(
+				"notification",
+				this,
+				CommandClasses["Multilevel Switch"],
+				pick(command, ["eventType", "direction"]),
```
Comment (AlCalzone) [inline packages/zwave-js/src/lib/node/Node.ts]:
```suggestion
				{ eventType: MultilevelSwitchCommand.StarLevelChange, direction: command.direction },
```

Comment (AlCalzone) [reply packages/zwave-js/src/lib/node/Node.ts]: (see above for explanation)

**Review by AlCalzone — CHANGES_REQUESTED** [review]
Sorry for taking so long to review this. A few more changes please.

File: packages/zwave-js/src/lib/commandclass/MultilevelSwitchCC.ts (hunk 12 above)
Comment (AlCalzone) [reply packages/zwave-js/src/lib/commandclass/MultilevelSwitchCC.ts]: Same goes for the other 2-3 lines where you reference this.

**Comment by reubenbijl** [comment]
@zwave-js-bot fix lint

**Review by AlCalzone — DISMISSED** [review]

**Review by AlCalzone — APPROVED** [review]

**Comment by AlCalzone** [comment]
Thanks a lot!

**Pull request merged**